### PR TITLE
Integration test for recreate masterkey command

### PR DESCRIPTION
**Pull request by sharidas**
Integration test for recreate masterkey command

Signed-off-by: Sujith H <sharidasan@owncloud.com>

## Description

Integration test for recreate masterkey occ command

## Motivation and Context

Integration test for recreate masterkey occ command.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: tests/integration/features/bootstrap/Provisioning.php [hunk 1]
```diff
@@ -107,6 +107,20 @@ public function creatingTheUser($user) {
 		$client->send($client->createRequest('GET', $url, $options2));
 	}
 
+	/**
+	 * @Given /^user "([^"]*)" reloggedin$/
+	 * @param $user
+	 */
+	public function loginAsUser($user) {
+		$password = ($user === 'admin') ? 'admin' : '123456';
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/Provisioning.php]: Hardcoded passwords are not good. There is a function to get them 

https://github.com/owncloud/core/blob/master/tests/integration/features/bootstrap/BasicStructure.php#L460

File: tests/integration/features/bootstrap/Provisioning.php [hunk 2]
```diff
@@ -107,6 +107,20 @@ public function creatingTheUser($user) {
...
+		$options = [
+			'auth' => [$user, $password]
+		];
+		$url = $this->baseUrl . "v{$this->apiVersion}.php/cloud/users" . '/' . $user;
+		$client = new Client();
+		$client->send($client->createRequest('GET', $url, $options));
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/Provisioning.php]: No exception catching here? the request could fail. I would only leave it this way if there is no possible situation when we need to continue the tests after this has failed.

File: tests/integration/features/recreatemasterkey.feature [hunk 3]
```diff
@@ -0,0 +1,23 @@
+Feature: recreate-master-key
+
+	@masterkey_encryption
+	Scenario: recreate masterkey
+		Given user "admin" exists
+		And user "admin" uploads file "data/textfile.txt" to "/somefile.txt"
+		When recreate masterkey by deleting old one and encrypting the filesystem
+		And the command was successful
+		Given user "admin" reloggedin
```
Comment (SergioBertolinSG) [inline tests/integration/features/recreatemasterkey.feature]: This scenario is confusing. There must be only one 'When', before it all steps has to be 'Given' steps and after it, all 'Then's.

File: tests/integration/features/recreatemasterkey.feature [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+		And as an "admin"
+		Then downloaded content when downloading file "/somefile.txt" with range "bytes=0-6" should be "This is"
+
+	@masterkey_encryption
+	Scenario: recreate masterkey and upload data
+		Given user "user0" exists
+		And user "user0" uploads file "data/textfile.txt" to "/somefile.txt"
+		When recreate masterkey by deleting old one and encrypting the filesystem
+		And the command was successful
+		Given user "admin" reloggedin
```
Comment (SergioBertolinSG) [inline tests/integration/features/recreatemasterkey.feature]: Same confusion here. Please always follow this rule: several Given, 1 When (thing to test), several Then.

File: tests/integration/features/bootstrap/CommandLine.php [hunk 5]
```diff
@@ -212,6 +212,15 @@ public function transferringOwnership($user1, $user2) {
 	}
 
 	/**
+	 * @When /^recreate masterkey by deleting old one and encrypting the filesystem/
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/CommandLine.php]: When recreating.

File: tests/integration/features/bootstrap/Provisioning.php [hunk 6]
```diff
@@ -107,6 +107,20 @@ public function creatingTheUser($user) {
 		$client->send($client->createRequest('GET', $url, $options2));
 	}
 
+	/**
+	 * @Given /^user "([^"]*)" reloggedin$/
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/Provisioning.php]: Not sure about the need of this function. It is just getting an url. Probably there are already functions to do this.

Even if there is a need for this function, the description isn't good. It should be given user blabla has relogged in.

**Review by SergioBertolinSG — CHANGES_REQUESTED** [review]

File: tests/integration/features/recreatemasterkey.feature (hunk 3 above)
Comment (sharidas) [reply tests/integration/features/recreatemasterkey.feature]: Updated both scenarios, to have "Given" before "When" and "Then" after when.

File: tests/integration/features/bootstrap/Provisioning.php (hunk 1 above)
Comment (sharidas) [reply tests/integration/features/bootstrap/Provisioning.php]: Removed this function. We don't need it. `logging in using web as user` is sufficient. Idea is login should happen after the re-creation of keys.

File: tests/integration/features/bootstrap/CommandLine.php (hunk 5 above)
Comment (sharidas) [reply tests/integration/features/bootstrap/CommandLine.php]: Updated to recreating

File: tests/integration/features/recreatemasterkey.feature [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+	@masterkey_encryption
+	Scenario: recreate masterkey and upload data
+		Given user "user0" exists
+		And user "user0" uploads file "data/textfile.txt" to "/somefile.txt"
+		When recreating masterkey by deleting old one and encrypting the filesystem
+		And the command was successful
+		And logging in using web as "admin"
+		And logging in using web as "user0"
+		And as an "user0"
+		And user "user0" uploads chunk file "1" of "1" with "AA" to "/somefile.txt"
```
Comment (SergioBertolinSG) [inline tests/integration/features/recreatemasterkey.feature]: THis is the when step. Previous are Given and next steps are Then (the first one Then, rest And)

File: tests/integration/features/recreatemasterkey.feature [hunk 8]
```diff
@@ -0,0 +1,24 @@
+Feature: recreate-master-key
+
+	@masterkey_encryption
+	Scenario: recreate masterkey
+		Given user "admin" exists
+		And user "admin" uploads file "data/textfile.txt" to "/somefile.txt"
+		When recreating masterkey by deleting old one and encrypting the filesystem
```
Comment (SergioBertolinSG) [inline tests/integration/features/recreatemasterkey.feature]: If recreating the masterkey is the actual test, the following steps should be Then. (1 Then, rest Ands).

**Review by SergioBertolinSG — CHANGES_REQUESTED** [review]

File: tests/integration/features/recreatemasterkey.feature [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+	@masterkey_encryption
+	Scenario: recreate masterkey and upload data
+		Given user "user0" exists
+		And user "user0" uploads file "data/textfile.txt" to "/somefile.txt"
+		When recreating masterkey by deleting old one and encrypting the filesystem
+		Then the command was successful
+		And logging in using web as "admin"
+		And logging in using web as "user0"
+		And as an "user0"
+		And user "user0" uploads chunk file "1" of "1" with "AA" to "/somefile.txt"
```
Comment (SergioBertolinSG) [inline tests/integration/features/recreatemasterkey.feature]: It looks like this step is the actual test. Recreating master key is just a precondition in this scenario.

**Review by SergioBertolinSG — CHANGES_REQUESTED** [review]

File: tests/integration/features/recreatemasterkey.feature [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+	@masterkey_encryption
+	Scenario: recreate masterkey
+		Given user "admin" exists
+		And user "admin" uploads file "data/textfile.txt" to "/somefile.txt"
+		And recreating masterkey by deleting old one and encrypting the filesystem
+		And the command was successful
+		And logging in using web as "admin"
+		And logging in using web as "admin"
+		When as an "admin"
```
Comment (sharidas) [inline tests/integration/features/recreatemasterkey.feature]: Looks ok for this scenario?

File: tests/integration/features/recreatemasterkey.feature [hunk 11]
```diff
@@ -0,0 +1,23 @@
+Feature: recreate-master-key
+
+	@masterkey_encryption
+	Scenario: recreate masterkey
+		Given user "admin" exists
+		And user "admin" uploads file "data/textfile.txt" to "/somefile.txt"
+		And recreating masterkey by deleting old one and encrypting the filesystem
+		And the command was successful
+		And logging in using web as "admin"
+		When as an "admin"
```
Comment (SergioBertolinSG) [inline tests/integration/features/recreatemasterkey.feature]: This was good in previous commit. 

When recreating masterkey by deleting old one and encrypting the filesystem

**Review by SergioBertolinSG — CHANGES_REQUESTED** [review]

**Review by SergioBertolinSG — APPROVED** [review]
If CI passes 👍

**Pull request merged**